src/persist-types/src/schema.rs

```diff
@@ -0,0 +1,431 @@
+                ));
+                *fields = f.finish().fields;
+            }
+        }
+    }
+}
+
+fn backward_compatible_typ(old: &DataType, new: &DataType) -> Option<ArrayMigration> {
+    use ArrayMigration::NoOp;
+    use DataType::*;
+    match (old, new) {
+        (Null, Null) => Some(NoOp),
+        (Null, _) => None,
+        (Boolean, Boolean) => Some(NoOp),
+        (Boolean, _) => None,
+        (Int8, Int8) => Some(NoOp),
+        (Int8, _) => None,
+        (Int16, Int16) => Some(NoOp),
+        (Int16, _) => None,
+        (Int32, Int32) => Some(NoOp),
+        (Int32, _) => None,
+        (Int64, Int64) => Some(NoOp),
+        (Int64, _) => None,
+        (UInt8, UInt8) => Some(NoOp),
+        (UInt8, _) => None,
+        (UInt16, UInt16) => Some(NoOp),
+        (UInt16, _) => None,
+        (UInt32, UInt32) => Some(NoOp),
+        (UInt32, _) => None,
+        (UInt64, UInt64) => Some(NoOp),
+        (UInt64, _) => None,
+        (Float16, Float16) => Some(NoOp),
+        (Float16, _) => None,
+        (Float32, Float32) => Some(NoOp),
+        (Float32, _) => None,
+        (Float64, Float64) => Some(NoOp),
+        (Float64, _) => None,
+        (Binary, Binary) => Some(NoOp),
+        (Binary, _) => None,
+        (FixedSizeBinary(o), FixedSizeBinary(n)) => (o == n).then_some(NoOp),
+        (FixedSizeBinary(_), _) => None,
+        (Utf8, Utf8) => Some(NoOp),
+        (Utf8, _) => None,
+        (Struct(o), Struct(n)) => backward_compatible_struct(o, n),
+        (Struct(_), _) => None,
+        (List(o), List(n)) => (o == n).then_some(NoOp),
+        (List(_), _) => None,
+        (Map(o, _), Map(n, _)) => (o == n).then_some(NoOp),
+        (Map(_, _), _) => None,
+        (
+            Timestamp(_, _)
+            | Date32
+            | Date64
+            | Time32(_)
+            | Time64(_)
+            | Duration(_)
+            | Interval(_)
+            | LargeBinary
+            | BinaryView
+            | LargeUtf8
+            | Utf8View
+            | ListView(_)
+            | FixedSizeList(_, _)
+            | LargeList(_)
+            | LargeListView(_)
+            | Union(_, _)
+            | Dictionary(_, _)
+            | Decimal128(_, _)
+            | Decimal256(_, _)
+            | RunEndEncoded(_, _),
+            _,
+        ) => unimplemented!("not used in mz: old={:?} new={:?}", old, new),
+    }
+}
+
+fn backward_compatible_struct(old: &Fields, new: &Fields) -> Option<ArrayMigration> {
+    use itertools::EitherOrBoth::*;
+    use ArrayMigration::*;
+    use StructArrayMigration::*;
+
+    // The common case is that fields will be in the same order, so start by
+    // assuming that and match them up pairwise. If this assumption is
+    // incorrect, then later there will be an add and drop for the same field
+    // name and we'll detect it then.
+    let old_new_fields = old
+        .iter()
+        .merge_join_by(new.iter(), |o, n| o.name().cmp(n.name()));
```

Comment:
Fixed. Don't love the result but it's semi-obvious and I think it works

Reply:
> FWIW I don't think we need drop support from the beginning for projection pushdown, we only don't register schemas when creating a ReadHandle so we might be okay here.

We'd still use this codepath to know how to actually project the data on read. It would just be handed an unregistered schema.

> I could be missing something, but it seems like a BTreeMap<ColumnName, (Option<Old>, Option<New>)> might do the trick here?

Yup, but I'd love to avoid the allocs. I think in all the real cases we'll be hitting the N^2 is better than the alloc. I considered some threshold-based fallback to `BTreeMap` (e.g. if either of them have at least N fields) but that codepath would basically be untested?

Reply:
Here's another case I'm not sure what to do with:

```
        testcase(
            struct_([("a", Boolean, true), ("c", Boolean, true)]),
            struct_([
                ("a", Boolean, true),
                ("b", Boolean, true),
                ("c", Boolean, true),
            ]),
            Some(false),
        );
```

This fails because `"b"` gets tacked on at the end instead of inserted in the middle - the migration struct doesn't have any way to express the position at which the insert takes place.

I don't think we expect to encounter this in practice, at least soon; it doesn't look super difficult to support, but we could also check for & disallow it or add a comment for future work.

Reply:
> I'm not sure I agree with "it doesn't look super difficult to support".

My concrete thought here was that this might be less error-prone if we changed the framing of the struct migration slightly - instead of "a list of changes to make to the input array to get the output", we could do "a list of the fields to include in the output, plus the (optional) index of the array in the input struct" or whatever.

I haven't actually tried it! But it seems like it would make at least make this kind of field-ordering bug much harder to write.

> Any suggestions on how to sniff it out?

No interesting ones! Seems like it should be detectable as we're looping through the fields... update a boolean when we first hit a field that's missing from the input, and then error if any of the remaining fields _aren't_ missing. But it does feel a bit delecate.